### ISSUE TITLE
Not able to send the data to iot-hub (HTTP|MQTT|AMQP)

### ISSUE THREAD

**Issue by nmohture** [Azure/iot-edge#396]
Hi ,

I am following ubuntu snappy on dell 3001 edge gw procedure https://github.com/Azure/iot-edge/blob/master/doc/azure_iot_gateway_snap.md and trying to run simulated device to cloud example, I have am facing issues for all the transport protocol, please find the error logs :

Logs
Transport: HTTP :
**************************
Press return to exit the application.
Device: 01:01:01:01:01:01, Temperature: 10.00
Device: 02:02:02:02:02:02, Temperature: 10.00
Error: Time:Tue Oct  3 05:50:36 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/c-utility/adapters/httpapi_curl.c Func:HTTPAPI_ExecuteRequest Line:594 Failure in HTTP communication: server reply code is 401
Info: HTTP Response:{"Message":"ErrorCode:IotHubUnauthorizedAccess;Unauthorized","ExceptionMessage":"Tracking ID:7c1b236b5fbe4df5812620eba72d0d02-G:6-TimeStamp:10/03/2017 00:20:36"}
***************************

Transport :MQTT:
***************************

(classic)admin@GTXQB02:~/TSE/parts/iot-edge/build$ ./samples/simulated_device_cloud_upload/simulated_device_cloud_upload_sample ../src/samples/simulated_device_cloud_upload/src/simulated_device_cloud_upload_lin.json
Press return to exit the application.
Device: 01:01:01:01:01:01, Temperature: 10.00
Device: 02:02:02:02:02:02, Temperature: 10.00
Error: Time:Tue Oct  3 05:52:22 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/c-utility/src/base64.c Func:Base64_Decoder Line:201 Invalid length Base64 string!
Error: Time:Tue Oct  3 05:52:22 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/c-utility/src/sastoken.c Func:construct_sas_token Line:211 Unable to decode the key for generating the SAS.
Error: Time:Tue Oct  3 05:52:22 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/iot-sdk-c/iothub_client/src/iothub_client_authorization.c Func:IoTHubClient_Auth_Get_SasToken Line:226 Failed creating sas_token
Error: Time:Tue Oct  3 05:52:22 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:1659 failure getting sas Token.
Error: Time:Tue Oct  3 05:52:22 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c Func:mqtt_operation_complete_callback Line:1397 Connection Not Accepted: 0x5: Not Authorized
Error: Time:Tue Oct  3 05:52:22 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/c-utility/src/tlsio_openssl.c Func:decode_ssl_received_bytes Line:695 SSL channel closed in decode_ssl_received_bytes.
Error: Time:Tue Oct  3 05:52:22 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/c-utility/src/tlsio_openssl.c Func:on_underlying_io_bytes_received Line:752 Error in decode_ssl_received_bytes.
 **************************************************

Transport : AMQP : 

****************************************************
(classic)admin@GTXQB02:~/TSE/parts/iot-edge/build$ vim ../src/samples/simulated_device_cloud_upload/src/simulated_device_cloud_upload_lin.json               (classic)admin@GTXQB02:~/TSE/parts/iot-edge/build$ ./samples/simulated_device_cloud_upload/simulated_device_cloud_upload_sample ../src/samples/simulated_device_cloud_upload/src/simulated_device_cloud_upload_lin.json
Press return to exit the application.
Device: 01:01:01:01:01:01, Temperature: 10.00
Device: 02:02:02:02:02:02, Temperature: 10.00
Info: Retry policy set (5, timeout = 0)
Info: Retry policy set (5, timeout = 0)
Info: Transport state changed from AMQP_TRANSPORT_STATE_NOT_CONNECTED to AMQP_TRANSPORT_STATE_CONNECTING
Info: Transport state changed from AMQP_TRANSPORT_STATE_CONNECTING to AMQP_TRANSPORT_STATE_CONNECTED
Error: Time:Tue Oct  3 05:53:32 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/c-utility/src/base64.c Func:Base64_Decoder Line:201 Invalid length Base64 string!
Error: Time:Tue Oct  3 05:53:32 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/c-utility/src/sastoken.c Func:construct_sas_token Line:211 Unable to decode the key for generating the SAS.
Error: Time:Tue Oct  3 05:53:32 2017 File:/home/admin/TSE/parts/iot-edge/src/deps/iot-sdk-c/iothub_client/src/iothubtransport_amqp_cbs_auth.c Func:create_and_put_SAS_token_to_cbs Line:383 Failed creating a SAS token (SASToken_Create() failed)
 
**********************************************************

Please let us know if any addition configuration is required ?

**Comment by vnn0804** [comment]
Hi ,
Does anybody facing this issue ?

**Comment by damonbarry** [comment]
I see those failures when I don't add my device's credentials to the `simulated_device_cloud_upload_lin.json` file.

**Comment by vnn0804** [comment]
hello damonbarry,
we added all credentials properly like IoTHubName, IoTHubSuffix, deviceId and deviceKey. but same issue occured.

**Comment by damonbarry** [comment]
@vnn0804 Let's figure out whether this is related to Ubuntu Snappy or not. Can you try this same sample outside of a snap (something like desktop Ubuntu, but same network if possible)? If that works it will help narrow the scope of the problem.

**Comment by Armin-FB** [comment]
Hi @nmohture,

I am facing the same issue (...base64.c Func:Base64_Decoder Line:201 Invalid length Base64 string!") using a MQTT Connection while sending via RaspberryPi to the IoT Hub.

Have you solved it yet?

Thanks in advance!
Armin

**Comment by damonbarry** [comment]
No activity. Please reopen if you'd like to continue to investigate...

**Comment by simf1** [comment]
Did you resolve this issue?

When my IoTDevKit is connected to the wifi. It repeats a message no-stop (see below).

I fallowed the step-by-step Microsoft Procedure "GetStarted - Connect IoT DevKit AZ3166 to Azure IoT Hub"

I took the ConnectionString from this command:

az iot hub device-identity show-connection-string --hub-name YourIoTHubName --device-id MyNodeDevice --output table
I tried with the connection located in Iot Device Explorer... and same result.

And i used this connection string format:

HostName={YourIoTHubName}.azure-devices.net;DeviceId=MyNodeDevice;SharedAccessKey={YourSharedAccessKey}
This is the ERROR:

2019-08-28 13:22:04 ERROR: base64_utility.c (ln 201): Invalid length Base64 string! 2019-08-28 13:22:04 ERROR: sastoken.c (ln 217): Unable to decode the key for generating the SAS. 2019-08-28 13:22:04 ERROR: iothub_client_authorization.c (ln 460): Failed creating sas_token 2019-08-28 13:22:04 ERROR: iothubtransport_mqtt_common.c (ln 2169): failure getting sas token from IoTHubClient_Auth_Get_SasToken.

The screen of my IoTDevKit still with the IP Adress and Running... (under that, " > IoT Hub")

Did you have a solution for that? My connection string contains only 112 characters (with the Label "HostName=") And... all worked fine the first day that I tried. The next day... this issue appears.

version
DevKitSDK version: 1.6.2

Mico version: 31620002.049

mbed-os version: 5.4.3

mbed TLS version: 2.15.0

WIFI version: wl0: Sep 10 2014 11:28:46 version 5.90.230.10
![2019-08-28 09_27_04-Window](https://user-images.githubusercontent.com/54630506/63865767-1bbed980-c980-11e9-8d81-fe941c52842a.png)
![2019-08-28 09_21_43-COM5 - PuTTY](https://user-images.githubusercontent.com/54630506/63865768-1bbed980-c980-11e9-9ffd-95f535011759.png)
